src/librustc_resolve/lib.rs

```diff
@@ -2569,45 +2569,150 @@ impl<'a> Resolver<'a> {
+                    "a constructor is private if any of the fields is private",
+                );
+                err
+            } else {
+                mk_struct_span_error(false)
+            }
+        } else {
+            mk_struct_span_error(false)
+        };
+
+        err.span_label(ident.span, &format!("this {} is private", binding.res().descr()));
+
+        // Display the chain of re-exports through to the original def for cases where the
+        // `use` is private but the def is public.
+        let mut imported = false;
+        let mut binding = *binding;
+        loop {
+            match binding.kind {
+                NameBindingKind::Res(res, _is_macro_export) => {
+                    match (res, imported, binding.vis) {
+                        (Res::Def(_, def_id), true, ty::Visibility::Public) => {
+                            // FIXME: we should verify that this def is actually
+                            // reachable from the user's crate, as the parent modules
+                            // of this ADT might be private.
+                            if def_id.is_local() {
+                                err.span_label(
+                                    binding.span,
+                                    &format!(
+                                        "consider importing {} `{}` directly",
+                                        res.descr(),
+                                        self.definitions
+                                            .def_path(def_id.index)
+                                            .to_string_no_crate(),
+                                    ),
+                                );
+                            } else {
+                                err.span_label(
+                                    binding.span,
+                                    &format!(
+                                        "consider importing {} `{}` directly",
+                                        res.descr(),
+                                        ident.name
+                                    ),
+                                );
+                            }
+                        }
+                        (Res::Def(_, def_id), true, _) if def_id.is_local() => {
+                            err.span_label(
+                                        binding.span,
+                                        &format!(
+                                            "consider making {} `{}` public",
+                                            res.descr(),
+                                            ident.name,
+                                        ),
+                                    );
+                        }
+                        _ => {
+                            err.span_label(
+                                binding.span,
+                                format!(
+                                    "{}{} `{}` declared here",
```

Comment:
```suggestion
                                    "{}{} `{}` is declared here",
```

Reply:
The PR uses the standard style right now -- I would prefer sticking to a uniform style. If we should change, let's do that in a PR dedicated to that.

Reply:
I'll open a PR to change the other occurrences, and then we can use "is" here from the start 😄

Reply:
Alright; r? r=me on that PR.